### PR TITLE
OpenSSL-1.1.0 asynchronous SSL/TLS mode support

### PR DESCRIPTION
**Pull request by mrpre**
OpenSSL provide asynchronous SSL/TLS mode since 1.1.0 version. This commit improve Tengine SSL layer to support the asynchronous mode based on OpenSSL-1.1.0f release. Docs and basic test case are also provided.

### CODE REVIEW

File: src/core/ngx_cycle.h [hunk 1]
```diff
@@ -60,6 +60,9 @@ struct ngx_cycle_s {
     ngx_connection_t         *connections;
     ngx_event_t              *read_events;
     ngx_event_t              *write_events;
+#if (NGX_SSL)
+    ngx_event_t              *async_events;
```
Comment (jinjiu) [inline src/core/ngx_cycle.h]: why no OPENSSL_VERSION_NUMBER

File: src/core/ngx_connection.c [hunk 2]
```diff
@@ -887,7 +887,16 @@ ngx_close_listening_sockets(ngx_cycle_t *cycle)
                      * for closed shared listening sockets unless
                      * the events was explicitly deleted
                      */
-
+#if (NGX_SSL)
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
```
Comment (jinjiu) [inline src/core/ngx_connection.c]: I suggest replacing it with "#if (NGX_SSL) && OPENSSL_VERSION_NUMBER >= 0x10100000L"

File: src/core/ngx_cycle.h (hunk 1 above)
Comment (mrpre) [reply src/core/ngx_cycle.h]: In ngx_core.h
```
#include <ngx_cycle.h>
...
#include <ngx_event_openssl.h>

```
When precompile the 'ngx_cycle.h', the macro 'OPENSSL_VERSION_NUMBER' has not been prepared, so it's meanless to add the macro to here.    

Also, the number of objects defined by ngx_cycle_t is limited, i think it's acceptable to add a element into ngx_cycle_t directly.

**Comment by chobits** [comment]
Two advice here:

1. should add C macro in nginx core, such as NGX_SSL_ASYNC, which makes merging nginx core easy
2. should add shell script to check whether ssl-async is supported during configure phase. If checking passes, enable NGX_SSL_ASYNC

**Comment by mrpre** [comment]
@chobits 
It's really difficult to detect whether the ssl library support the async mode.   
It all depends on the version of OpenSSL and the implement of engine library in 'LIBPATH/engines-1.1/'. 
  
  

Shoud I just simply replace '#if (NGX_SSL) && OPENSSL_VERSION_NUMBER >= 0x10100000L' by '#if NGX_SSL_ASYNC' and whether the  'NGX_SSL_ASYNC' is defined depends on the OpenSSL version and 'NGX_SSL' during configure phase?

**Comment by chobits** [comment]
太多的原生逻辑修正，请务必不要考虑安全性和原先代码逻辑bugfix，只添加与ssl async相关逻辑的修正。

There are too many changes to modify nginx core, pls make ssl async logic clear & easy to review

File: src/core/ngx_connection.c (hunk 2 above)
Comment (mrpre) [reply src/core/ngx_connection.c]: fixed

File: src/core/ngx_cycle.h (hunk 1 above)
Comment (mrpre) [reply src/core/ngx_cycle.h]: fixed

File: src/event/ngx_event_openssl.c [hunk 3]
```diff
@@ -1222,6 +1246,82 @@ ngx_ssl_set_session(ngx_connection_t *c, ngx_ssl_session_t *session)
     return NGX_OK;
 }
 
+
+#if (NGX_SSL && NGX_SSL_ASYNC)
+ngx_int_t
+ngx_ssl_async_process_fds(ngx_connection_t *c)
```
Comment (jinjiu) [inline src/event/ngx_event_openssl.c]: Why do not use NGX_OK and NGX_ERROR, but with 0 and 1

File: src/event/ngx_event_openssl.c [hunk 4]
```diff
@@ -4357,6 +4745,10 @@ ngx_openssl_engine(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
 
     oscf->engine = 1;
 
+    if(cf->no_ssl_init) {
```
Comment (jinjiu) [inline src/event/ngx_event_openssl.c]: if( --> if (

File: src/http/modules/ngx_http_ssl_module.c [hunk 5]
```diff
@@ -857,6 +888,38 @@ ngx_http_ssl_enable(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
...
 
+#if (NGX_HTTP_SSL && NGX_SSL_ASYNC)
+static char *
+ngx_http_ssl_enable_async(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
+{
+    ngx_http_ssl_srv_conf_t *sscf = conf;
+
+    char  *rv;
+
+    ngx_flag_t       *pssl, *pssl_async;
```
Comment (jinjiu) [inline src/http/modules/ngx_http_ssl_module.c]: Variable declaration format should be aligned

File: src/http/ngx_http_request.c [hunk 6]
```diff
@@ -1408,12 +1420,21 @@ ngx_http_read_request_header(ngx_http_request_t *r)
         return n;
     }
 
-    if (rev->ready) {
+#if (NGX_HTTP_SSL && NGX_SSL_ASYNC)
+    if(c->async_enable)
         n = c->recv(c, r->header_in->last,
+               r->header_in->end - r->header_in->last);
```
Comment (jinjiu) [inline src/http/ngx_http_request.c]: Wrap format should be aligned

File: src/http/ngx_http_request.c [hunk 7]
```diff
@@ -1408,12 +1420,21 @@ ngx_http_read_request_header(ngx_http_request_t *r)
...
+    else {
+#endif
+        if (rev->ready) {
+            n = c->recv(c, r->header_in->last,
                     r->header_in->end - r->header_in->last);
```
Comment (jinjiu) [inline src/http/ngx_http_request.c]: Wrap format should be aligned

File: src/event/ngx_event_openssl.c (hunk 3 above)
Comment (mrpre) [reply src/event/ngx_event_openssl.c]: Done

File: docs/modules/ngx_http_ssl_asynchronous_mode_cn.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+Description
+===========
+
+本文档提供关于如何在Nginx开启异步SSL/TLS支持的说明.
+* 异步SSL/TLS模式是OpenSSL 1.1.0版本之后引入的新的模式
+
+编译支持
+===========
+
+启--with-openssl-async用编译选项
```
Comment (jinjiu) [inline docs/modules/ngx_http_ssl_asynchronous_mode_cn.md]: image
应该是：
启用--with-openssl-async编译选项

**Pull request merged**